File: hasPalindromePermutation.js
function hasPalindromePermutation(theString) {
  // declare object to hold characters and count of each in string
  let charCount = {};
  
  // loop thru string length to add to object
  for (let i = 0; i < theString.length; i++) {
    let currentChar = theString[i];
    
    if (charCount[theString[i]]) {
      charCount[currentChar]++
    } else {
      charCount[currentChar] = 1
    }
  }
  
  // checking if any permutation is palindrome  
  let oddCount = 0; 
  
  for (let char in charCount) {
    if (charCount[char] % 2 !== 0) {
      oddCount++;
    }
  }
  
  if (oddCount > 1) {
    return false; 
  }

  return true;
}



/* 
Input string
Output boolean
Constraints: can't be an unordered set?
Edge cases: empty string, one character, same character odd number of times


ATTACK:
1. create an object representing characters string
2. loop thru object
3. create variable to hold "middle" character which can be odd (any variation of odd no.)
4. if multiple characters are odd ( > 1), then not palindrome


for each character in the string
check if word is odd or even
{
  c: 2, i: 2: v; 1
}

if one character in string- that's a palindrome
if even no. for all characters- that's a palindrome  
if odd no. of characs- then one character key on the object has to be odd (ie civvvic) 
civil = false c = 1, i = 2, v = 1, l = 1 civic = 1 odd character, civvvic = 3 odd = true
*/


// Tests

// let desc = 'permutation with odd number of chars';
// assertEqual(hasPalindromePermutation('aabcbcd'), true, desc);

// desc = 'permutation with even number of chars';
// assertEqual(hasPalindromePermutation('aabccbdd'), true, desc);

// desc = 'no permutation with odd number of chars';
// assertEqual(hasPalindromePermutation('aabcd'), false, desc);

// desc = 'no permutation with even number of chars';
// assertEqual(hasPalindromePermutation('aabbcd'), false, desc);

// desc = 'empty string';
// assertEqual(hasPalindromePermutation(''), true, desc);

// desc = 'one character string ';
// assertEqual(hasPalindromePermutation('a'), true, desc);

// function assertEqual(a, b, desc) {
//   if (a === b) {
//     console.log(`${desc} ... PASS`);
//   } else {
//     console.log(`${desc} ... FAIL: ${a} != ${b}`);
//   }
// }

File: hasPalindromePermutation.py
import unittest

# This is my initial implementation:
# # check if any permutation of a string is a palindrome
# # (a palindrome may have an odd amount of one character, all other characters must have an even count)
#
# def has_palindrome_permutation(string):
#
#     # Create a dict counting the amount of each character in the string
#     char_counts = {}
#     for ch in string:
#         char_counts[ch] = char_counts.get(ch, 0) + 1
#
#     # Check that there is no more than one character with an odd amount
#     has_odd = False
#     for count in char_counts.values():
#         if count % 2 != 0:
#             if has_odd:
#                 return False
#             has_odd = True
#
#     return True

# I implemented the second-best solution, according to InterviewCake. Tracking which characters have an odd count in a
# set, adding and removing them as you go, would a) not keep track of amounts you don't actually need to use,
# b) prevent integer overflow if you were dealing with a VERY large string, and c) doesn't require you to loop through
# the dictionary at the end to count odd numbers.
# Here's the better solution implemented

def has_palindrome_permutation(string):

    # Create a set containing each character with an odd amount
    odd_chars = set()
    for ch in string:
        if ch in odd_chars:
            odd_chars.discard(ch)
        else:
            odd_chars.add(ch)

    # Check whether there is no more than one character with an odd amount
    return not (len(odd_chars) > 1)


# # Tests
#
# class Test(unittest.TestCase):
#
#     def test_permutation_with_odd_number_of_chars(self):
#         result = has_palindrome_permutation('aabcbcd')
#         self.assertTrue(result)
#
#     def test_permutation_with_even_number_of_chars(self):
#         result = has_palindrome_permutation('aabccbdd')
#         self.assertTrue(result)
#
#     def test_no_permutation_with_odd_number_of_chars(self):
#         result = has_palindrome_permutation('aabcd')
#         self.assertFalse(result)
#
#     def test_no_permutation_with_even_number_of_chars(self):
#         result = has_palindrome_permutation('aabbcd')
#         self.assertFalse(result)
#
#     def test_empty_string(self):
#         result = has_palindrome_permutation('')
#         self.assertTrue(result)
#
#     def test_one_character_string(self):
#         result = has_palindrome_permutation('a')
#         self.assertTrue(result)
#
#
# unittest.main(verbosity=2)